DatadogSessionReplay/Sources/Recorder/RecordingCoordinator.swift

```diff
@@ -0,0 +1,42 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-Present Datadog, Inc.
+ */
+
+import Foundation
+import Datadog
+
+internal protocol RecordingCoordination {
+    var currentRUMContext: RUMContext? { get }
+}
+
+/// Object is responsible for getting the RUM context, randomising the sampling rate,
+/// starting/stopping the recording scheduler as needed and propagating `has_replay` to other features.
+internal class RecordingCoordinator: RecordingCoordination {
+    /// Last received RUM context (or `nil` if RUM session is not sampled).
+    /// It's synchronized through `scheduler.queue` (main thread).
+    private(set) var currentRUMContext: RUMContext? = nil
+
+    init(
+        scheduler: Scheduler,
+        rumContextObserver: RUMContextObserver,
+        contextPublisher: SRContextPublisher,
+        sampler: Sampler
+    ) {
+        contextPublisher.setRecordingIsPending(false)
+
+        rumContextObserver.observe(on: scheduler.queue) { [weak self] rumContext in
+            if self?.currentRUMContext?.ids.sessionID != rumContext?.ids.sessionID || self?.currentRUMContext == nil {
+                let isSampled = sampler.sample() && rumContext != nil
+                contextPublisher.setRecordingIsPending(isSampled)
+                if isSampled {
+                    scheduler.start()
+                } else {
+                    scheduler.stop()
+                }
+            }
+            self?.currentRUMContext = rumContext
+        }
```

Comment:
question/ The `scheduler` is not started initially and it will be started with the first change to RUM context. Will it work no matter of the order of enabling RUM and SR?

Imagine this situation:
```swift
RUM.enable()
RUMMonitor.shared().startView() // changes RUM context

SR.enable()
// ... recording is not started
// ... it will be only started with next change to RUM context
```
We seem to introduce logical race condition here, no? Ideally, we should be able to check the initial RUM context and perform initial `scheduler.start()`. This however is not yet doable in V2 - we lack an API to read current SDK context, we can only observe it and (I believe) observation doesn't emit initial value.

Reply:
Technically order matters. If RUM starts publishing on message bus before SR gets initialized it could miss some messages. I would say - it's by design currently. We could think about potential direct dependency between RUM and SR. Personally I'm in favour of decoupling with risk above in mind.

In reality it gets frequently updated, so we should "catch" RUM context fairly quickly regardless the order.